signer/core/signed_data.go

```diff
@@ -471,7 +472,7 @@ func (typedData *TypedData) EncodeData(primaryType string, data map[string]inter
 					if err != nil {
 						return nil, err
 					}
-					arrayBuffer.Write(encodedData)
+					arrayBuffer.Write(crypto.Keccak256(encodedData))
```

Comment:
Could you explain this change? 
It looks like something that was either completely broken earlier, or is completely broken now. If it's the former, does any test show how it was broken?

Reply:
### why `hash` before `write`
According to the describe on [EIP-712](https://eips.ethereum.org/EIPS/eip-712) , when `EncodeData` on a type of `array`, it need to `hash` the result of each element, and then append these results together, `hash` the combination again. 

It may look like this:

```python
sample = [{'k':'a', 'v':'b'}, {'k':'c', 'v':'d'}]
# old version
data = hash(hash(a) + hash(b) + hash(c) + hash(d))
# fixed
data = hash(hash(hash(a) + hash(b)) + hash(hash(c) + hash(d)))
```
_Curious_: I see [Line 494](https://github.com/ethereum/go-ethereum/blob/release%2F1.10/signer/core/signed_data.go#L494) correctly use 
```golang
buffer.Write(crypto.Keccak256(encodedData))
```
instead of 
```golang
buffer.Write(encodedData)
```
why not do the same in this line? It's the same condition (both are returned from deeper `EncodeData`)
### why the former test not broken
Reason is simple, cause the test datas in `signed_data_test.go` are not in full coverage. No array in any test samples
### more
I have done the test with MetaMask